File: src/main/java/me/testaccount666/migration/plugins/essentials/MuteMigrator.java
package me.testaccount666.migration.plugins.essentials;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.moderation.MuteModeration;
import me.testaccount666.serversystem.userdata.UserManager;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;

public class MuteMigrator extends AbstractMigrator {

    @Override
    public int migrateFrom() {
        var count = 0;

        var userManager = userManager();
        var essentials = essentials();
        var uuids = essentials.getUsers().getAllUserUUIDs();

        for (var uuid : uuids) {
            var userOptional = userManager.getUser(uuid);
            if (userOptional.isEmpty()) {
                ServerSystem.getLog().warning("Couldn't find user '${uuid}', skipping state migration!");
                continue;
            }

            var essentialsUser = essentials.getUser(uuid);
            var user = userOptional.get().getOfflineUser();

            if (!essentialsUser.isMuted()) continue;

            var defaultReason = command("Moderation.DefaultReason", UserManager.Companion.getConsoleUser())
                    .target(essentialsUser.getName()).prefix(false).send(false).build();
            if (defaultReason.isEmpty()) {
                ServerSystem.getLog().severe("(MuteMigrator) Default reason is empty! This should not happen! (Mute Migration cancelled)");
                return count;
            }

            var muteManager = user.getMuteManager();
            var expireTime = essentialsUser.getMuteTimeout();
            var issueTime = System.currentTimeMillis(); // Issue time is lost

            var reason = essentialsUser.getMuteReason();
            if (reason == null) reason = defaultReason.get();

            var senderUUID = UserManager.Companion.getConsoleUser().getUuid(); // Sender UUID is lost
            var targetUUID = user.getUuid();

            muteManager.addModeration(new MuteModeration(false, issueTime, expireTime, reason, senderUUID, targetUUID));
            user.save();

            count += 1;
        }

        return count;
    }

    @Override
    public int migrateTo() {
        var count = 0;

        var userManager = userManager();
        var essentials = essentials();

        for (var player : offlinePlayers()) {
            var uuid = player.getUniqueId();

            var userOptional = userManager.getUser(uuid);
            if (userOptional.isEmpty()) {
                ServerSystem.getLog().warning("Couldn't find user '${uuid}', skipping mute migration!");
                continue;
            }

            var user = userOptional.get().getOfflineUser();

            ensureUserDataExists(uuid);
            var essentialsUser = essentials.getUser(uuid);

            var muteManager = user.getMuteManager();
            var muteOptional = muteManager.getActiveModeration();
            if (muteOptional.isEmpty()) continue;

            var mute = muteOptional.get();
            essentialsUser.setMuted(true);
            essentialsUser.setMuteReason(mute.reason());
            essentialsUser.setMuteTimeout(mute.expireTime());
            // Again, `issueTime` and `senderUUID` is lost

            count += 1;
        }

        return count;
    }
}

File: src/main/kotlin/me/testaccount666/migration/plugins/essentials/MuteMigrator.kt
package me.testaccount666.migration.plugins.essentials

import me.testaccount666.serversystem.ServerSystem.Companion.log
import me.testaccount666.serversystem.extensions.commandMsg
import me.testaccount666.serversystem.moderation.MuteModeration
import me.testaccount666.serversystem.userdata.UserManager.Companion.consoleUser
import java.util.logging.Level

class MuteMigrator : AbstractMigrator() {
    override fun migrateFrom(): Int {
        return essentials.users.allUserUUIDs.count { uuid ->
            runCatching {
                val cachedUser = userManager.getUserOrNull(uuid) ?: run {
                    log.warning("Couldn't find user '${uuid}', skipping state migration!")
                    return@count false
                }

                val essentialsUser = essentials.getUser(uuid)
                val user = cachedUser.offlineUser

                if (!essentialsUser.isMuted) return@count false

                val defaultReason = consoleUser.commandMsg("Moderation.DefaultReason") {
                    target(essentialsUser.name)
                    prefix(false)
                    send(false)
                    blankError(true)
                }

                if (defaultReason.isEmpty()) {
                    log.severe("(MuteMigrator) Default reason is empty! This should not happen!")
                    return@count false
                }

                val muteManager = user.muteManager
                val expireTime = essentialsUser.muteTimeout
                val issueTime = System.currentTimeMillis() // Issue time is lost

                val reason = essentialsUser.muteReason ?: defaultReason

                val senderUUID = consoleUser.uuid // Sender UUID is lost
                val targetUUID = user.uuid

                muteManager.addModeration(
                    MuteModeration.builder {
                        isShadowMute(false)
                        targetUuid(targetUUID)
                        issueTime(issueTime)
                        expireTime(expireTime)
                        senderUuid(senderUUID)
                        reason(reason)
                    }
                )
                user.save()

                return@runCatching true
            }.onFailure { log.log(Level.WARNING, "Couldn't migrate mute for '${uuid}'", it) }.getOrDefault(false)
        }
    }

    override fun migrateTo(): Int {
        return offlinePlayers().count { player ->
            val uuid = player.uniqueId

            runCatching {
                val cachedUser = userManager.getUserOrNull(uuid) ?: run {
                    log.warning("Couldn't find user '${uuid}', skipping mute migration!")
                    return@count false
                }

                val user = cachedUser.offlineUser

                ensureUserDataExists(uuid)
                val essentialsUser = essentials.getUser(uuid)

                val muteManager = user.muteManager
                val mute = muteManager.activeModeration ?: return@count false

                essentialsUser.muted = true
                essentialsUser.muteReason = mute.reason
                essentialsUser.muteTimeout = mute.expireTime

                // Again, `issueTime` and `senderUUID` is lost
                return@runCatching true
            }.onFailure { log.log(Level.WARNING, "Couldn't migrate mute for '${uuid}'", it) }.getOrDefault(false)
        }
    }
}
